R/nm-join.R

```diff
@@ -0,0 +1,179 @@
+#' input data set and any tables you want to join. We recommend you make this
+#' column a simple integer numbering the rows in the input data set (for example
+#' `NUM`). When this column is carried into the output table files, there will
+#' be unambiguous matching from the table file back to the input data set.
+#'
+#' The one exception to this are `$FIRSTONLY` tables. If a table file has the
+#' same number of rows as the there are individuals in the input data set, it
+#' will assumed to be a `$FIRSTONLY` table. In this case, the table will be
+#' joined to the input data by the `ID` column. If `ID` is not present in the
+#' table, it will be using `.join_col`. Note that if _neither_ `ID` or the
+#' column passed to `.join_col` are present in the table, the join will fail.
+#'
+#' Note also that, when `.join_col` is carried into table outputs, **there is no
+#' need to table any other columns from the input data** as long as the
+#' `nm_join()` approach is used; any column in the input data set, regardless
+#' of whether it is listed in `$INPUT` or not, will be carried through from the
+#' input data and therefore available in the joined result.
+#'
+#' **Duplicate columns are dropped**
+#'
+#' If a table has columns with the same name as columns in the input data set,
+#' or a table that has already been joined, those columns will be dropped from
+#' the joined data. If `getOption(bbr.verbose) == TRUE` a message will be
+#' printed about any columns dropped this way.
+#'
+#' The one exception to this is the `DV` column. If `DV` is present in the input
+#' data _and_ at least one of the table files, the `DV` column from the input
+#' data will be renamed to `DV.DATA` and the column from the table file kept as
+#' `DV`.
+#'
+#' @importFrom dplyr left_join select
+#' @seealso [nm_tables()], [nm_table_files()], [nm_file()]
+#' @export
+nm_join <- function(
+  .mod,
+  .join_col = "NUM",
+  .files = nm_table_files(.mod),
+  .superset = FALSE,
+  .bbi_args = list(
+    no_grd_file = TRUE,
+    no_shk_file = TRUE
+  )
+) {
+
+  if (inherits(.mod, "character")) {
+    checkmate::assert_string(.mod)
+    .mod <- read_model(.mod)
+  }
+  check_model_object(.mod, c(NM_MOD_CLASS, NM_SUM_CLASS))
+  checkmate::assert_string(.join_col)
+  checkmate::assert_character(.files)
+  checkmate::assert_logical(.superset, len = 1)
+  checkmate::assert_list(.bbi_args)
+
+  df_list <- nm_tables(.mod, .files = .files)
+  .d <- df_list$data
+  .tbls <- df_list[2:length(df_list)]
+  if (
+    "DV" %in% names(.d) &&
+    "DV" %in% unlist(map(.tbls, names))
+  ) {
+    .d <- rename(.d, DV.DATA = .data$DV)
+  }
+  col_order <- names(.d)
+
+  .join_col <- toupper(.join_col)
+  if (!(.join_col %in% names(.d))) {
+    stop(glue("couldn't find `.join_col` {.join_col} in data with cols: {paste(names(.d), collapse = ', ')}"))
+  }
+
+  if (.superset) {
+    join_fun <- function(x, y, ...) left_join(y, x, ...)
```

Comment:
Is it just a stylistic choice to swap the `left_join` args rather than use `right_join`?  (The tests pass for me if I use `right_join`, so I think it might be.) If it makes a practical difference, it might be worth a comment/test case to avoid a regression by someone "simplifying" it later.

Reply:
I think it had something to do with the column ordering, but since we're doing that manually at the bottom then this shouldn't matter at all. I agree: let's switch it to prevent confusion down the road.